packages/block-editor/src/components/block-switcher/index.js

```diff
@@ -12,252 +12,154 @@ import {
+			const styles = firstBlock && getBlockStyles( firstBlock.name );
+			return {
+				blocks: _blocks,
+				inserterItems: getInserterItems( rootClientId ),
+				hasBlockStyles: !! styles?.length,
+			};
+		},
+		[ clientIds ]
 	);
-}
-
-export class BlockSwitcher extends Component {
-	constructor() {
-		super( ...arguments );
-		this.state = {
-			hoveredClassName: null,
-		};
-		this.onHoverClassName = this.onHoverClassName.bind( this );
-	}
-
-	onHoverClassName( className ) {
-		this.setState( { hoveredClassName: className } );
-	}
-
-	render() {
-		const {
-			blocks,
-			onTransform,
-			inserterItems,
-			hasBlockStyles,
-		} = this.props;
-		const { hoveredClassName } = this.state;
-
-		if ( ! Array.isArray( blocks ) || ! blocks.length ) {
-			return null;
-		}
-
-		const [ hoveredBlock ] = blocks;
-		const itemsByName = mapKeys( inserterItems, ( { name } ) => name );
-		const possibleBlockTransformations = orderBy(
-			filter(
-				getPossibleBlockTransformations( blocks ),
-				( block ) => block && !! itemsByName[ block.name ]
-			),
-			( block ) => itemsByName[ block.name ].frecency,
-			'desc'
-		);
-
-		// When selection consists of blocks of multiple types, display an
-		// appropriate icon to communicate the non-uniformity.
-		const isSelectionOfSameType =
-			uniq( map( blocks, 'name' ) ).length === 1;
-
-		let icon;
-		if ( isSelectionOfSameType ) {
-			const sourceBlockName = hoveredBlock.name;
-			const blockType = getBlockType( sourceBlockName );
-			icon = blockType.icon;
-		} else {
-			icon = stack;
-		}
 
-		const hasPossibleBlockTransformations = !! possibleBlockTransformations.length;
+	if ( ! blocks?.length ) return null;
 
-		if ( ! hasBlockStyles && ! hasPossibleBlockTransformations ) {
-			return (
-				<ToolbarGroup>
-					<ToolbarButton
-						disabled
-						className="block-editor-block-switcher__no-switcher-icon"
-						title={ __( 'Block icon' ) }
-						icon={ <BlockIcon icon={ icon } showColors /> }
-					/>
-				</ToolbarGroup>
-			);
-		}
-
-		const blockSwitcherLabel =
-			1 === blocks.length
-				? __( 'Change block type or style' )
-				: sprintf(
-						/* translators: %s: number of blocks. */
-						_n(
-							'Change type of %d block',
-							'Change type of %d blocks',
-							blocks.length
-						),
-						blocks.length
-				  );
+	const onTransform = ( name ) =>
+		replaceBlocks( clientIds, switchToBlockType( blocks, name ) );
 
+	const [ hoveredBlock ] = blocks;
+	// When selection consists of blocks of multiple types, display an
+	// appropriate icon to communicate the non-uniformity.
+	const isSelectionOfSameType =
+		uniq( blocks.map( ( { name } ) => name ) ).length === 1;
+	let icon;
+	if ( isSelectionOfSameType ) {
+		const sourceBlockName = hoveredBlock.name;
+		const blockType = getBlockType( sourceBlockName );
+		icon = blockType.icon;
+	} else {
+		icon = stack;
+	}
+	const itemsByName = mapKeys( inserterItems, ( { name } ) => name );
+	const possibleBlockTransformations = orderBy(
```

Comment:
This array is the only thing that's really needed in the component itself and things like inserterItems are not really needed. I wonder if it means we lack a direct selector for that. Relying on "inserterItems" which is a big selector in itself can be time consuming. (It was there before, just raising something I remarked)

Reply:
That's actually so right - we need much less information. I'll create a new selector for this.

Reply:
Made a first draft of a new selector here: https://github.com/WordPress/gutenberg/pull/27674/commits/4c3dfc70ea52810e86c843d5a3abc2d428d3b855 to have a first look @youknowriad :)